internal/pkg/template/app.go

```diff
@@ -0,0 +1,105 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package template
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+	"unicode"
+)
+
+// Names of application templates.
+const (
+	lbWebAppTemplateName = "lb-web-app"
+)
+
+// AppNestedStackOpts holds optional configuration that's needed only if the application stack has a nested stack.
+type AppNestedStackOpts struct {
+	StackName string
+
+	VariableOutputs []string
+	SecretOutputs   []string
+	PolicyOutputs   []string
+}
+
+// AppOpts holds optional data that can be provided to render an application stack template.
+// The empty AppOpts{} struct means that none of these fields will be rendered in the application template.
+type AppOpts struct {
+	Variables map[string]string
+	Secrets   map[string]string
+
+	// Outputs from nested stacks such as the addons stack.
+	NestedStack *AppNestedStackOpts
+}
+
+// LoadBalancedWebAppConfig holds data that needs be provided to render a load balanced web app stack template.
+type LoadBalancedWebAppConfig struct {
+	// Optional fields.
+	AppOpts
+
+	// Mandatory fields.
+	RulePriorityLambda string
+}
+
+// ParseLoadBalancedWebApp parses a load balanced web app's CloudFormation template
+// with the specified data object and returns its content.
+func (t *Template) ParseLoadBalancedWebApp(data LoadBalancedWebAppConfig) (*Content, error) {
+	return t.ParseApp(lbWebAppTemplateName, data, withAppParsingFuncs())
+}
```

Comment:
The load balanced stuff seems out of place here, do you think so? Otherwise this is a perfectly generic little file.

Reply:
Yeaaa. I'm really not sure so any alternatives you might have in mind is welcomed :D I wanted at least the following criteria:
- Like we discussed offline, `ParseApp` should be ideally avoided because it accepts an `interface{}` so it's possible that an error happens at runtime instead of compile time if the user doesn't provide an object with the required fields to render an app template.

My reasoning with these methods and structs are as follows:
1. To make sure the data passed in is valid at compile-time, we need to have structs or interfaces. Since the apps share data, I went with structs. Providing an empty `LoadBalancedWebAppConfig` guarantees the template will be rendered.
2. I thought that the structs needed to render an app template should be defined in the `template` pkg instead of the `stack` pkg because it's supposed to be the package that touches the template files.

Does that make sense? I'm not sure either :/ I don't hate this but I agree that it's also not nice mostly because we have to define new structs again.

Reply:
After an offline discussion, we decided to remove the `LoadBalancedWebAppConfig` and only use the `AppOpts` struct to represent the collection of features that would result in a change in the app cfn templates.